addons/website_sale/models/website.py

```diff
@@ -101,18 +119,21 @@ def get_pricelist_available(self, show_visible=False):
             else:
                 # In the weird case we are coming from the backend (https://github.com/odoo/odoo/issues/20245)
                 website = len(self) == 1 and self or self.search([], limit=1)
-        isocountry = request and request.session.geoip and request.session.geoip.get('country_code') or False
+        isocountry = req and req.session.geoip and req.session.geoip.get('country_code') or False
         partner = self.env.user.partner_id
         last_order_pl = partner.last_website_so_id.pricelist_id
         partner_pl = partner.sudo(user=self.env.user).property_product_pricelist
         pricelists = website._get_pl_partner_order(isocountry, show_visible,
                                                    website.user_id.sudo().partner_id.property_product_pricelist.id,
-                                                   request and request.session.get('website_sale_current_pl') or None,
+                                                   req and req.session.get('website_sale_current_pl') or None,
                                                    website.pricelist_ids,
                                                    partner_pl=partner_pl and partner_pl.id or None,
                                                    order_pl=last_order_pl and last_order_pl.id or None)
         return self.env['product.pricelist'].browse(pricelists)
 
+    def get_pricelist_available(self, show_visible=False):
```

Comment:
To remove in master?

Reply:
I encapsulated the `get_pricelist_available()` in `_get_pricelist_available()` to be able to easily mock the method.
So Odoo's code is still calling `get_pricelist_available()`, it should not be removed.

See this code in `test_website_sale_pricelist.py` that is in `[FIX] website_sale: fix _get_pl_partner_order` commit of this PR:
```python
class DotDict(dict):
    """dot.notation access to dictionary attributes"""
    def __getattr__(*args):
        val = dict.get(*args)
        return DotDict(val) if type(val) is dict else val

def get_pricelist_available(self, show_visible=False, website_id=1, country_code=None, website_sale_current_pl=None):
    request = DotDict({
        'website': self.env['website'].browse(website_id),
        'session': {
            'geoip': {
                'country_code': country_code,
            },
            'website_sale_current_pl': website_sale_current_pl,
        },
    })
    return self.env['website']._get_pricelist_available(request, show_visible)
```

I know it is a bit weird but to test `get_pricelist_available()` from python, it is the only way really to simulate a frontend context, as:
1. We don't want to test this method in tour for performance reason
2. We want to write lot of tests on that method as it is the core method of `website_sale`
3. We can't use `request` in tests as it is unbound at that time and `get_pricelist_available()` is trying to read `request` attributes (see [here](https://github.com/odoo/odoo/blame/d12c993a28b5971dc5bf7dbb88e4b7c606f592b5/addons/website_sale/models/website.py#L89), [here](https://github.com/odoo/odoo/blame/d12c993a28b5971dc5bf7dbb88e4b7c606f592b5/addons/website_sale/models/website.py#L96) and [here](https://github.com/odoo/odoo/blame/d12c993a28b5971dc5bf7dbb88e4b7c606f592b5/addons/website_sale/models/website.py#L102).)

It is the cleanest way I came up with to test `website_sale` behaviors regarding pricelists.
That way, Odoo's code still use `get_pricelist_available()` that call `_get_pricelist_available()` and pass `request` while tests call directly `_get_pricelist_available()` and give `DotDict` class as request parameter (that mock perfectly the regular request).

Reply:
ok